### PR TITLE
[SPARK-29894][SQL][WEBUI] Add Codegen Stage Id to Spark plan graphs in Web UI SQL Tab

### PR DESCRIPTION
**Pull request by LucaCanali**
### What changes were proposed in this pull request?
The Web UI SQL Tab provides information on the executed SQL using plan graphs and by reporting SQL execution plans. Both sources provide useful information. Physical execution plans report Codegen Stage Ids. This PR adds Codegen Stage Ids to the plan graphs.

### Why are the changes needed?
It is useful to have Codegen Stage Id information also reported in plan graphs, this allows to more easily match physical plans and graphs with metrics when troubleshooting SQL execution.
Example snippet to show the proposed change: 

Example of the current state: 

Physical plan:

### Does this PR introduce any user-facing change?
This PR adds Codegen Stage Id information to SQL plan graphs in the Web UI/SQL Tab.  

### How was this patch tested?
Added a test + manually tested

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SparkPlanGraph.scala [hunk 1]
```diff
@@ -189,18 +191,19 @@ private[ui] class SparkPlanGraphCluster(
     name: String,
     desc: String,
     val nodes: mutable.ArrayBuffer[SparkPlanGraphNode],
-    metrics: Seq[SQLPlanMetric])
-  extends SparkPlanGraphNode(id, name, desc, metrics) {
+    metrics: Seq[SQLPlanMetric],
+    codegenStageId: Option[Int] = None)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SparkPlanGraph.scala]: can we include the codegen id when building the `name`? Then we don't need to change the class signature here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SparkPlanGraph.scala [hunk 2]
```diff
@@ -88,7 +88,8 @@ object SparkPlanGraph {
           planInfo.nodeName,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SparkPlanGraph.scala]: We can construct the name with `codegenStageId` here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala [hunk 3]
```diff
@@ -71,6 +77,8 @@ private[execution] object SparkPlanInfo {
       plan.nodeName,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala]: We can construct the name with `codegenStageId` here.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala]: then the diff can be much smaller.

Comment (LucaCanali) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala]: Thanks for reviwing. The proposal sounds good to me.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SparkPlanGraph.scala [hunk 4]
```diff
@@ -77,8 +77,9 @@ object SparkPlanGraph {
       parent: SparkPlanGraphNode,
       subgraph: SparkPlanGraphCluster,
       exchanges: mutable.HashMap[SparkPlanInfo, SparkPlanGraphNode]): Unit = {
+    val WholeStageCodegenPattern = "(WholeStageCodegen.*)".r
     planInfo.nodeName match {
-      case "WholeStageCodegen" =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SparkPlanGraph.scala]: how about `case name if name.startsWith("WholeStageCodegen")`

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
@LucaCanali we need to update org.apache.spark.sql.execution.metric.SQLMetricsSuite.WholeStageCodegen metrics

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**